test/src/test-capi-array-write-ordered-attr-fixed.cc

```diff
@@ -0,0 +1,405 @@
+    std::vector<T> output_data(expected_data.size());
+
+    // Create subarray.
+    tiledb_subarray_t* subarray;
+    require_tiledb_ok(tiledb_subarray_alloc(ctx, array, &subarray));
+    require_tiledb_ok(tiledb_subarray_add_range(
+        ctx, subarray, 0, &min_index, &max_index, nullptr));
+
+    // Create write query.
+    tiledb_query_t* query;
+    require_tiledb_ok(tiledb_query_alloc(ctx, array, TILEDB_READ, &query));
+    require_tiledb_ok(tiledb_query_set_layout(ctx, query, TILEDB_ROW_MAJOR));
+    require_tiledb_ok(tiledb_query_set_subarray_t(ctx, query, subarray));
+    require_tiledb_ok(tiledb_query_set_data_buffer(
+        ctx, query, "a", output_data.data(), &attr_data_size));
+
+    // Submit read query and verify it is successful.
+    check_tiledb_ok(tiledb_query_submit(ctx, query));
+    tiledb_query_status_t query_status;
+    check_tiledb_ok(tiledb_query_get_status(ctx, query, &query_status));
+    CHECK(query_status == TILEDB_COMPLETED);
+
+    CHECK(output_data == expected_data);
+
+    // Clean-up.
+    tiledb_subarray_free(&subarray);
+    tiledb_query_free(&query);
+    tiledb_array_free(&array);
+  }
+
+  /**
+   * Write data to the ordered attribute.
+   *
+   * @param min_index Lower bound for range to write data to.
+   * @param max_index Upper bound for range to write data to.
+   * @param data Data to write to the attribute.
+   * @param valid If ``true`` the write should succeed, otherwise the write
+   *     should fail.
+   */
+  void write_fragment(
+      uint32_t min_index,
+      uint32_t max_index,
+      std::vector<T>& data,
+      bool valid = true) {
+    tiledb_array_t* array;
+    require_tiledb_ok(tiledb_array_alloc(ctx, array_name.c_str(), &array));
+    require_tiledb_ok(tiledb_array_open(ctx, array, TILEDB_WRITE));
+
+    // Define sizes for setting buffers.
+    uint64_t attr_data_size{data.size() * sizeof(T)};
+
+    // Create subarray.
+    tiledb_subarray_t* subarray;
+    require_tiledb_ok(tiledb_subarray_alloc(ctx, array, &subarray));
+    require_tiledb_ok(tiledb_subarray_add_range(
+        ctx, subarray, 0, &min_index, &max_index, nullptr));
+
+    // Create write query.
+    tiledb_query_t* query;
+    require_tiledb_ok(tiledb_query_alloc(ctx, array, TILEDB_WRITE, &query));
+    require_tiledb_ok(tiledb_query_set_layout(ctx, query, TILEDB_ROW_MAJOR));
+    require_tiledb_ok(tiledb_query_set_subarray_t(ctx, query, subarray));
+    require_tiledb_ok(tiledb_query_set_data_buffer(
+        ctx, query, "a", data.data(), &attr_data_size));
+
+    // Submit query.
+    if (valid) {
+      // Submit write query and verify it is successful.
+      check_tiledb_ok(tiledb_query_submit(ctx, query));
+      tiledb_query_status_t query_status;
+      check_tiledb_ok(tiledb_query_get_status(ctx, query, &query_status));
+      CHECK(query_status == TILEDB_COMPLETED);
+    } else {
+      // Submit write query and verify if fails.
+      auto rc = tiledb_query_submit(ctx, query);
```

Comment:
Can we check that we are hitting the expected error? We have lots of tests that don't really test what they say they are testing.

Reply:
Same for the file below.